tool/tsh/common/vnet_common.go

```diff
@@ -49,10 +72,105 @@ func (p *vnetAppProvider) GetProfile(profileName string) (*profile.Profile, erro
+// leaf clusters of that root.
+func (p *vnetAppProvider) ReissueAppCert(ctx context.Context, profileName, leafClusterName string, app types.Application) (tls.Certificate, error) {
+	tc, err := p.newTeleportClient(ctx, profileName, leafClusterName)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	var cert tls.Certificate
+	err = p.retryWithRelogin(ctx, tc, func() error {
+		var err error
+		cert, err = p.reissueAppCert(ctx, tc, profileName, leafClusterName, app)
+		return trace.Wrap(err, "reissuing app cert")
+	})
+	return cert, trace.Wrap(err)
+}
+
+// GetRootClusterCACertPool returns a certificate pool for the root cluster of the given profile.
+func (p *vnetAppProvider) GetRootClusterCACertPool(ctx context.Context, profileName string) (*x509.CertPool, error) {
+	tc, err := p.newTeleportClient(ctx, profileName, "")
+	if err != nil {
+		return nil, trace.Wrap(err, "creating new client")
+	}
+	certPool, err := tc.RootClusterCACertPool(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading root cluster CA cert pool")
+	}
+	return certPool, nil
+}
+
+func (p *vnetAppProvider) retryWithRelogin(ctx context.Context, tc *client.TeleportClient, fn func() error, opts ...client.RetryWithReloginOption) error {
+	profileName, err := utils.Host(tc.WebProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	opts = append(opts,
+		client.WithBeforeLoginHook(func() error {
+			fmt.Printf("Login for cluster %s expired, attempting to log in again.\n", tc.SiteName)
+			return nil
+		}),
+		client.WithAfterLoginHook(func() error {
+			return trace.Wrap(p.clientCache.ClearForRoot(profileName), "clearing client cache after relogin")
+		}),
+	)
+	return client.RetryWithRelogin(ctx, tc, fn, opts...)
+}
+
+func (p *vnetAppProvider) reissueAppCert(ctx context.Context, tc *client.TeleportClient, profileName, leafClusterName string, app types.Application) (tls.Certificate, error) {
+	slog.InfoContext(ctx, "Reissuing cert for app.", "app_name", app.GetName(), "profile", profileName, "leaf_cluster", leafClusterName)
+
+	routeToApp := proto.RouteToApp{
+		Name:        app.GetName(),
+		PublicAddr:  app.GetPublicAddr(),
+		ClusterName: tc.SiteName,
+	}
+
+	profile, err := tc.ProfileStatus()
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err, "loading client profile")
+	}
+
+	appCertParams := client.ReissueParams{
+		RouteToCluster: tc.SiteName,
+		RouteToApp:     routeToApp,
+		AccessRequests: profile.ActiveRequests.AccessRequests,
+	}
+
+	clusterClient, err := p.GetCachedClient(ctx, profileName, leafClusterName)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err, "getting cached cluster client")
+	}
+	rootClient, err := p.GetCachedClient(ctx, profileName, "")
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err, "getting cached root client")
+	}
+
+	key, err := appLogin(ctx, tc, clusterClient, rootClient.AuthClient, appCertParams)
```

Comment:
Is it necessary to pass `rootClient.AuthClient` here? Can we just use `clusterClient.AuthClient` inside?

p.s. I suppose this is also a question for myself, as I'm the culprit...

Reply:
clusterClient.AuthClient doesn't work because that will call to the leaf cluster auth, instead of the root, and it fails fetching the user while trying to create the app session

Reply:
and i'd rather not create the root client inside with `clusterClient.ConnectToRootCluster`, because in vnet i am using cached clients for all clusters